clippy_lints/src/methods/clone_on_arc_or_rc.rs

```diff
@@ -0,0 +1,57 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::msrvs::{self, Msrv};
+use clippy_utils::source::snippet;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_ast::ast::UnOp;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_span::{Symbol, sym};
+
+use super::CLONE_ON_ARC_OR_RC;
+
+pub(super) fn check(
+    cx: &LateContext<'_>,
+    expr: &Expr<'_>,
+    method_name: Symbol,
+    receiver: &Expr<'_>,
+    args: &[Expr<'_>],
+    msrv: &Msrv,
+) {
+    if !msrv.meets(msrvs::ARC_RC_UNWRAP_OR_CLONE) {
+        return;
+    }
+
+    if method_name == sym::clone
+        && args.is_empty()
+        && let ExprKind::Unary(UnOp::Deref, recv) = receiver.kind
+        && let Some(arc_or_rc_path) = is_arc_or_rc(cx, recv)
+    {
+        span_lint_and_sugg(
+            cx,
+            CLONE_ON_ARC_OR_RC,
+            expr.span,
+            "using `unwrap_or_clone` on an `Arc` or `Rc` is more efficient",
+            "try",
+            format!(
+                "{arc_or_rc_path}::unwrap_or_clone({snip})",
+                snip = snippet(cx, recv.span, "..")
+            ),
+            Applicability::MachineApplicable,
+        );
+    }
+}
+
+#[allow(clippy::similar_names)]
+fn is_arc_or_rc(cx: &LateContext<'_>, expr: &Expr<'_>) -> Option<&'static str> {
+    let is_arc = is_type_diagnostic_item(cx, cx.typeck_results().node_type(expr.hir_id), sym::Arc);
+    let is_rc = is_type_diagnostic_item(cx, cx.typeck_results().node_type(expr.hir_id), sym::Rc);
+
+    if is_arc {
+        Some("std::sync::Arc")
+    } else if is_rc {
+        Some("std::rc::Rc")
+    } else {
+        None
+    }
```

Comment:
Maybe you could use something like

```suggestion
    match get_type_diagnostic_name(cx, cx.typeck_results().expr_ty(expr)) {
        Some(sym::Arc) => Some("std::sync::Arc"),
        Some(sym::Rc) => Some("std::rc::Rc"),
        _ => None
    }
```

instead and avoid the multiple lookups.

Reply:
I didn't know `get_type_diagnostic_name`. Fixed it in https://github.com/rust-lang/rust-clippy/pull/13921/commits/0986e73b4da44f2438fc20449da610cee4fc305d. Thanks!